### DISCUSSION TITLE
Configurable storageProviderTimeout

### DISCUSSION THREAD

**Discussion by dasniko** [keycloak/keycloak#11989] [Archive]
Hi @hmlnarik 

this topic is not directly about the NEW store, but a store issue...

I just stumbled upon an old, not resolved, issue here: [KEYCLOAK-18856](https://issues.redhat.com/browse/KEYCLOAK-18856)
The _not-so-configurable_ `storageProviderTimeout`.

A customer of mine has just this requirement to increase the current "fixed" 3000ms timeout.
Is there any way to achieve it?

For now, I only have a (very dirty) reflection approach:

Configuring a custom `storageProviderTimeout` property value for the custom SPI.
Then, use this code in the `CustomStorageProviderFactory`:

```
private Long storageProviderTimeout = 3000L;

@Override
public void init(Config.Scope config) {
    storageProviderTimeout = config.getLong("storageProviderTimeout", storageProviderTimeout);
}

@Override
public void postInit(KeycloakSessionFactory factory) {
    KeycloakSession session = factory.create();
    UserProvider userProvider = session.userStorageManager();
    try {
        Field field = userProvider.getClass().getSuperclass().getDeclaredField("STORAGE_PROVIDER_DEFAULT_TIMEOUT");
        field.setAccessible(true);
        Field modifiers = field.getClass().getDeclaredField("modifiers");
        modifiers.setAccessible(true);
        modifiers.setInt(field, field.getModifiers() & ~Modifier.FINAL);
        field.set(userProvider, storageProviderTimeout);
    } catch (NoSuchFieldException | IllegalAccessException e) {
        throw new RuntimeException(e);
    }
}
```
(seen [here](https://keycloak.discourse.group/t/how-to-configure-storageprovidertimeout/9171/10?u=dasniko)) 
But I don't feel very comfortable with it. ;)

is there any change to get a fix in one of the next versions? Or will it only be fixed with the new store and new provider management, and thus it will still take some time?

Thanks & regards,
\- Niko

**Comment by hmlnarik** [comment]
@dasniko Thank you for the report.
@mhajas Could you please chime in?

**Comment by mhajas** [comment]
Hello @dasniko, if I understand correctly (please correct me if I am wrong), the only issue is that we are not able to set the value for  `storageProviderTimeout` in `standalone.xml` because the schema doesn't allow it. This is present only in Wildfly distribution using the legacy store. Since it is the legacy store and legacy distribution, I don't think we have any plans to address this in the next release. On the other hand, the fix should be pretty straightforward; just allowing the property to be set in the correct place so that `AbstractStorageManager` can pick it up. Am I missing something?

**Comment by thomasdarimont** [comment]
The following setting works for me with Keycloak 18.0.0 on Quarkus `--spi-user-storage-provider-timeout=2000`

**Reply by nkuruza** [reply]
How can I add the same setting for the Wildfly distribution?

**Reply by mhajas** [reply]
@nkuruza I am afraid it is not possible to set with the Wildfly distribution.

**Reply by nkuruza** [reply]
Thank you very much for the response, @mhajas. I guess I will have try the dirty reflection approach provided above. 

Thanks again.

**Reply by mhajas** [reply]
Consider switching to Quarkus distribution. New releases won't contain Wildfly dist anymore.

**Reply by mhagnumdw** [reply]
> The following setting works for me with Keycloak 18.0.0 on Quarkus `--spi-user-storage-provider-timeout=2000`

Hi! Does the `--spi-user-storage-provider-timeout` parameter exist in the documentation? Where?!

**Comment by dasniko** [comment]
Thanks @mhajas and @thomasdarimont for responding.
I wasn‘t aware of the fact, that it really is „only“ a schema issue in Wildfly. I formerly read the KEYCLOAK-18856 issue in the way that it was a wrong implementation in the `AbstractUserManager`, but it seems that it is working as expected in Quarkus, as @thomasdarimont already mentioned and I confirm this. Unfortunately I didn‘t had time to check this earlier.

The initial request, when creating this discussion threat, I had to solve it somehow for the legacy Wildfly distribution. But I don‘t think it has to be fixed for Wildfly. One reason more to convince the customer to migrate better sooner than later. :)
